### PR TITLE
[FLINK-4865] [ml] Add EvaluateDataSet operation for LabeledVector

### PR DESCRIPTION
**Pull request by tfournier314**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by thvasilo** [comment]
Thank you for your contribution Thomas!

The changes look mostly fine I think, my only concern is the need to cast the types in lines [215-216](https://github.com/apache/flink/pull/2684/files#diff-c74d49c94a86f962939a42362f3ea190R215), I'm not sure if there is a more elegant way to do that.

I'm going to ping @chiwanpark, @mbalassi and @tillrohrmann, if anyone has some time to review and merge this let us know in the comments.

As a side note, I've noticed that we don't include the evaluate operation in the documentation of all Predictors, so I've created [FLINK-4908](https://issues.apache.org/jira/browse/FLINK-4908) to track that.

**Comment by tfournier314** [comment]
Ok thanks !

We don't need to care about Jenkins/Travis fails, do we ?

**Comment by thvasilo** [comment]
> We don't need to care about Jenkins/Travis fails, do we ?

Normally we should, but there many unstable tests and issues with Travis that it's hard to follow now.
You should generally make sure the failures don't originate from code you have contributed, and then it's up to the commiter to decide.

**Comment by thvasilo** [comment]
One last thing, could you change this PR's title to `[FLINK-4865] [ml]  Add EvaluateDataSet operation for LabeledVector`

This is the standard naming for PRs, issue number, followed by component, followed by the description. That way the PR gets linked back to the JIRA issue automatically.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @tfournier314. 

I'm wondering whether we should introduce a new `EvaluateDataSetOperation` for the special case of a `LabeledVector`. Can't you easily achieve the same by `val evaluateDS = labeledVectorDS.map(x => (x.vector, x.label))` and then using `evaluateDS` for the `evaluate` operation?

In general I think it would be better to write a generic `EvaluateDataSetOperation` which can take a `ValueExtractor` which can extract the value from the `Testing` type. Then we could offer a `ValueExtractor` for the `LabeledVector`. That way we could extend it easily for different types as well.

A general remark: It's always helpful to write a PR description and format the title according to the contribution guidelines [1]. Furthermore, your code does not adhere to the existing code style. Even though Flink does not have a strict code style, it is always good to stick to the code style of the existing code.

[1] http://flink.apache.org/how-to-contribute.html

**Comment by thvasilo** [comment]
Hello @tillrohrmann I think you make some valid points. My original idea was to move completely to using `LabeledVector` to the de facto format for supervised learning problems and do away with `(Vector, Double)` tuples.

What you suggest would add some extra flexibility though, esp. in terms of `(FeatureVector, LabelVector)` tuples which we have discussed in the context of ANNs.

@tfournier314 Could you investigate the applicability of the scheme that Till recommended?

**Comment by tfournier314** [comment]
@thvasilo @tillrohrmann Ok I will investigate this

**Comment by zentol** [comment]
Closing since flink-ml is effectively frozen.
